application/common/models/EmailQueue.php

```diff
@@ -0,0 +1,219 @@
+<?php
+namespace common\models;
+
+use common\helpers\Utils;
+use common\models\User;
+use yii\helpers\ArrayHelper;
+use yii\helpers\Json;
+
+class EmailQueue extends EmailQueueBase
+{
+    public function rules()
+    {
+        return ArrayHelper::merge(
+            [
+                [
+                    'attempts_count', 'default', 'value' => 0,
+                ],
+                [
+                    'created', 'default', 'value' => Utils::getDatetime(),
+                ],
+                [
+                    ['to_address', 'cc_address'], 'email',
+                ],
+                [
+                    'event_log_user_id', 'exist', 'targetClass' => User::className(),
+                    'targetAttribute' => ['event_log_user_id' => 'id'],
+                ],
+            ],
+            parent::rules()
+        );
+    }
+
+    /**
+     * Attempt to send an email, but on error queue it.
+     * Throws exception if send and queue fail
+     * @param string $toAddress
+     * @param string $subject
+     * @param string $textBody
+     * @param null|string $htmlBody
+     * @param null|string $ccAddress
+     * @param null|integer $eventLogUserId
+     * @param null|string $eventLogTopic
+     * @param null|string $eventLogDetails
+     * @return EmailQueue
+     * @throws \Exception
+     */
+    public static function sendOrQueue(
+        $toAddress,
+        $subject,
+        $textBody,
+        $htmlBody = null,
+        $ccAddress = null,
+        $eventLogUserId = null,
+        $eventLogTopic = null,
+        $eventLogDetails = null
+    ) {
+        $emailQueue = new EmailQueue();
+        $emailQueue->to_address = $toAddress;
+        $emailQueue->subject = $subject;
+        $emailQueue->text_body = $textBody;
+        $emailQueue->html_body = $htmlBody;
+        $emailQueue->cc_address = $ccAddress;
+        $emailQueue->event_log_user_id = $eventLogUserId;
+        $emailQueue->event_log_topic = $eventLogTopic;
+        $emailQueue->event_log_details = $eventLogDetails;
+
+        $emailQueue->send();
+
+        return $emailQueue;
```

Comment:
hmm, I have the queue in send on a failure, I suppose it could be elevated to here if an exception occurs during send so that send is explicitly only sending, not sure if it matters though